File: ts/middleware.ts
/**
 * X402 Express Middleware
 * Protect your API endpoints with 402 Payment Required responses.
 */

import type { Request, Response, NextFunction, RequestHandler } from 'express';
import type { X402Headers } from './types';
import { X402Client } from './client';

export interface PaymentConfig {
  amount: number;
  currency?: 'SOL' | 'USDC';
  recipient: string;
  expiresIn?: number;
  memo?: string;
}

export interface MiddlewareConfig {
  apiKey: string;
  baseUrl?: string;
  recipient: string;
  verifyPayments?: boolean;
  onPaymentReceived?: (payment: PaymentInfo, req: Request) => void | Promise<void>;
  onPaymentFailed?: (error: Error, req: Request) => void | Promise<void>;
}

export interface PaymentInfo {
  signature: string;
  reference: string;
  amount: number;
  currency: string;
  sender?: string;
  verifiedAt: Date;
}

const usedReferences = new Set<string>();

export function createX402Middleware(config: MiddlewareConfig) {
  const client = new X402Client({
    apiKey: config.apiKey,
    baseUrl: config.baseUrl,
  });

  function requirePayment(amount: number, options: Partial<PaymentConfig> = {}): RequestHandler {
    return async (req: Request, res: Response, next: NextFunction) => {
      const paymentSignature = req.headers['x-402-payment-signature'] as string;
      const paymentReference = req.headers['x-402-payment-reference'] as string;

      if (paymentSignature && paymentReference) {
        if (usedReferences.has(paymentReference)) {
          return res.status(402).json({
            error: 'Payment reference already used',
            code: 'REPLAY_ATTACK',
          });
        }

        if (config.verifyPayments !== false) {
          try {
            const verification = await client.verify({
              signature: paymentSignature,
              reference: paymentReference,
              expectedAmount: amount,
              expectedRecipient: options.recipient || config.recipient,
            });

            if (!verification.verified) {
              if (config.onPaymentFailed) {
                await config.onPaymentFailed(new Error('Payment verification failed'), req);
              }
              return res.status(402).json({
                error: 'Payment verification failed',
                code: 'VERIFICATION_FAILED',
              });
            }

            usedReferences.add(paymentReference);

            const paymentInfo: PaymentInfo = {
              signature: paymentSignature,
              reference: paymentReference,
              amount: verification.transaction.amount,
              currency: verification.transaction.currency,
              sender: verification.transaction.sender,
              verifiedAt: new Date(),
            };

            (req as any).x402Payment = paymentInfo;

            if (config.onPaymentReceived) {
              await config.onPaymentReceived(paymentInfo, req);
            }

            return next();
          } catch (error) {
            if (config.onPaymentFailed) {
              await config.onPaymentFailed(error as Error, req);
            }
            return res.status(402).json({
              error: 'Payment verification error',
              code: 'VERIFICATION_ERROR',
              message: (error as Error).message,
            });
          }
        } else {
          usedReferences.add(paymentReference);
          (req as any).x402Payment = {
            signature: paymentSignature,
            reference: paymentReference,
            amount,
            currency: options.currency || 'SOL',
            verifiedAt: new Date(),
          };
          return next();
        }
      }

      const reference = generateReference();
      const expiresIn = options.expiresIn || 600;
      const expires = Math.floor(Date.now() / 1000) + expiresIn;

      const headers: X402Headers = {
        'X-402-Version': '1.0',
        'X-402-Amount': amount.toString(),
        'X-402-Currency': options.currency || 'SOL',
        'X-402-Recipient': options.recipient || config.recipient,
        'X-402-Reference': reference,
        'X-402-Expires': expires.toString(),
      };

      Object.entries(headers).forEach(([key, value]) => {
        res.setHeader(key, value);
      });

      return res.status(402).json({
        error: 'Payment Required',
        code: 'PAYMENT_REQUIRED',
        payment: {
          amount,
          currency: options.currency || 'SOL',
          recipient: options.recipient || config.recipient,
          reference,
          expires,
          memo: options.memo,
        },
        instructions: 'Send payment to the recipient address and retry with X-402-Payment-Signature and X-402-Payment-Reference headers',
      });
    };
  }

  function optionalPayment(amount: number, options: Partial<PaymentConfig> = {}): RequestHandler {
    return async (req: Request, res: Response, next: NextFunction) => {
      const paymentSignature = req.headers['x-402-payment-signature'] as string;
      
      if (paymentSignature) {
        return requirePayment(amount, options)(req, res, next);
      }

      (req as any).x402Payment = null;
      (req as any).x402PaymentAvailable = {
        amount,
        currency: options.currency || 'SOL',
        recipient: options.recipient || config.recipient,
      };
      
      return next();
    };
  }

  return {
    requirePayment,
    optionalPayment,
    client,
  };
}

function generateReference(): string {
  const timestamp = Date.now().toString(36);
  const random = Math.random().toString(36).substring(2, 15);
  return `pay_${timestamp}${random}`;
}

export function getPaymentInfo(req: Request): PaymentInfo | null {
  return (req as any).x402Payment || null;
}

export function isPaid(req: Request): boolean {
  return !!(req as any).x402Payment;
}

File: python/x402_sdk/middleware.py
"""
X402 Flask Middleware
Protect your API endpoints with 402 Payment Required responses.
"""

import time
import functools
from typing import Optional, Callable, Any, Dict, Set
from datetime import datetime

from flask import request, jsonify, Response

from .client import X402Client
from .types import PaymentInfo


_used_references: Set[str] = set()


class X402Middleware:
    """Flask middleware for X402 payment protection."""
    
    def __init__(
        self,
        api_key: str,
        recipient: str,
        base_url: str = "https://api.x402agent.tech",
        verify_payments: bool = True,
        on_payment_received: Optional[Callable[[PaymentInfo], None]] = None,
        on_payment_failed: Optional[Callable[[Exception], None]] = None,
    ):
        self.client = X402Client(api_key=api_key, base_url=base_url)
        self.recipient = recipient
        self.verify_payments = verify_payments
        self.on_payment_received = on_payment_received
        self.on_payment_failed = on_payment_failed
    
    def require_payment(
        self,
        amount: float,
        currency: str = "SOL",
        recipient: Optional[str] = None,
        expires_in: int = 600,
        memo: Optional[str] = None,
    ) -> Callable:
        """Decorator to require payment for an endpoint."""
        
        def decorator(f: Callable) -> Callable:
            @functools.wraps(f)
            def wrapped(*args, **kwargs):
                payment_signature = request.headers.get("X-402-Payment-Signature")
                payment_reference = request.headers.get("X-402-Payment-Reference")
                
                actual_recipient = recipient or self.recipient
                
                if payment_signature and payment_reference:
                    if payment_reference in _used_references:
                        return jsonify({
                            "error": "Payment reference already used",
                            "code": "REPLAY_ATTACK",
                        }), 402
                    
                    if self.verify_payments:
                        try:
                            from .types import VerifyRequest
                            verification = self.client.verify(VerifyRequest(
                                signature=payment_signature,
                                reference=payment_reference,
                                expected_amount=amount,
                                expected_recipient=actual_recipient,
                            ))
                            
                            if not verification.verified:
                                if self.on_payment_failed:
                                    self.on_payment_failed(Exception("Verification failed"))
                                return jsonify({
                                    "error": "Payment verification failed",
                                    "code": "VERIFICATION_FAILED",
                                }), 402
                            
                            _used_references.add(payment_reference)
                            
                            payment_info = PaymentInfo(
                                signature=payment_signature,
                                reference=payment_reference,
                                amount=verification.transaction.amount,
                                currency=verification.transaction.currency,
                                sender=verification.transaction.sender,
                                verified_at=datetime.now(),
                            )
                            
                            request.x402_payment = payment_info
                            
                            if self.on_payment_received:
                                self.on_payment_received(payment_info)
                            
                            return f(*args, **kwargs)
                            
                        except Exception as e:
                            if self.on_payment_failed:
                                self.on_payment_failed(e)
                            return jsonify({
                                "error": "Payment verification error",
                                "code": "VERIFICATION_ERROR",
                                "message": str(e),
                            }), 402
                    else:
                        _used_references.add(payment_reference)
                        request.x402_payment = PaymentInfo(
                            signature=payment_signature,
                            reference=payment_reference,
                            amount=amount,
                            currency=currency,
                            verified_at=datetime.now(),
                        )
                        return f(*args, **kwargs)
                
                reference = _generate_reference()
                expires = int(time.time()) + expires_in
                
                response = jsonify({
                    "error": "Payment Required",
                    "code": "PAYMENT_REQUIRED",
                    "payment": {
                        "amount": amount,
                        "currency": currency,
                        "recipient": actual_recipient,
                        "reference": reference,
                        "expires": expires,
                        "memo": memo,
                    },
                    "instructions": "Send payment to the recipient address and retry with X-402-Payment-Signature and X-402-Payment-Reference headers",
                })
                response.status_code = 402
                response.headers["X-402-Version"] = "1.0"
                response.headers["X-402-Amount"] = str(amount)
                response.headers["X-402-Currency"] = currency
                response.headers["X-402-Recipient"] = actual_recipient
                response.headers["X-402-Reference"] = reference
                response.headers["X-402-Expires"] = str(expires)
                
                return response
            
            return wrapped
        return decorator


def create_x402_middleware(
    api_key: str,
    recipient: str,
    base_url: str = "https://api.x402agent.tech",
    verify_payments: bool = True,
) -> X402Middleware:
    """Create an X402 middleware instance."""
    return X402Middleware(
        api_key=api_key,
        recipient=recipient,
        base_url=base_url,
        verify_payments=verify_payments,
    )


def require_payment(
    amount: float,
    currency: str = "SOL",
    recipient: Optional[str] = None,
    expires_in: int = 600,
) -> Callable:
    """Standalone decorator for requiring payment (requires middleware to be initialized)."""
    def decorator(f: Callable) -> Callable:
        @functools.wraps(f)
        def wrapped(*args, **kwargs):
            reference = _generate_reference()
            expires = int(time.time()) + expires_in
            
            payment_signature = request.headers.get("X-402-Payment-Signature")
            
            if payment_signature:
                return f(*args, **kwargs)
            
            response = jsonify({
                "error": "Payment Required",
                "code": "PAYMENT_REQUIRED",
                "payment": {
                    "amount": amount,
                    "currency": currency,
                    "recipient": recipient or "RECIPIENT_ADDRESS",
                    "reference": reference,
                    "expires": expires,
                },
            })
            response.status_code = 402
            response.headers["X-402-Version"] = "1.0"
            response.headers["X-402-Amount"] = str(amount)
            response.headers["X-402-Currency"] = currency
            response.headers["X-402-Reference"] = reference
            response.headers["X-402-Expires"] = str(expires)
            
            return response
        
        return wrapped
    return decorator


def _generate_reference() -> str:
    """Generate a unique payment reference."""
    import random
    import string
    timestamp = hex(int(time.time()))[2:]
    random_part = ''.join(random.choices(string.ascii_lowercase + string.digits, k=12))
    return f"pay_{timestamp}{random_part}"


def get_payment_info() -> Optional[PaymentInfo]:
    """Get payment info from the current request."""
    return getattr(request, 'x402_payment', None)


def is_paid() -> bool:
    """Check if the current request has a valid payment."""
    return hasattr(request, 'x402_payment') and request.x402_payment is not None
